### DISCUSSION TITLE
exploring-dotnet-6-part-8-improving-logging-performance-with-source-generators/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#12] [Announcements]
# exploring-dotnet-6-part-8-improving-logging-performance-with-source-generators/

In this post I look at the subtleties of using ILogger and discuss the new source generator that can give performance benefits without the boilerplate

https://andrewlock.net/exploring-dotnet-6-part-8-improving-logging-performance-with-source-generators/

**Comment by septst** [comment]
This is such a brilliant article! I have been using SeriLog for my structured logging needs and never been aware of all these different ways, especially how the filtered-out logs add to the memory allocation. I made a quick check and Serilog seems to do the ```IsEnabled``` check for LogLevels. Thank you!

**Comment by MaxKot** [comment]
Thank you for your article.

However, the benchmark is misleading.

Your `InMemoryLogger` never calls the `formatter` delegate passed to it so it never actually turns the template string `"Writing hello world response to {Person}"` into the message `"Writing hello world response to Person { Id = 123, Name = Joe Blogs }"`. While some log output formats do not format the message, a lot of them do. I believe writing the formatted message should be added to the benchmark or at least a statement should be made that the benchmark only reflects performance of writing unformatted message templates only.

I got the following benchmark results after adding an option to invoke the `formatter` delegate:

|                              Method | InvokeFormatter |       Mean |     Error |    StdDev |  Gen 0 | Allocated |
|------------------------------------ |---------------- |-----------:|----------:|----------:|-------:|----------:|
|                 **InterpolatedEnabled** |           **False** | **258.327 ns** | **4.4419 ns** | **4.1550 ns** | **0.0014** |     **664 B** |
|                InterpolatedDisabled |           False | 253.641 ns | 2.7430 ns | 2.5658 ns | 0.0014 |     664 B |
|                   DirectCallEnabled |           False |  41.694 ns | 0.5299 ns | 0.4956 ns | 0.0001 |      64 B |
|                  DirectCallDisabled |           False |  36.625 ns | 0.3777 ns | 0.3348 ns | 0.0001 |      64 B |
|               IsEnabledCheckEnabled |           False |  45.380 ns | 0.4842 ns | 0.4529 ns | 0.0001 |      64 B |
|              IsEnabledCheckDisabled |           False |   3.552 ns | 0.0533 ns | 0.0473 ns |      - |         - |
|          LoggerMessageDefineEnabled |           False |  22.692 ns | 0.4632 ns | 0.5858 ns |      - |         - |
|         LoggerMessageDefineDisabled |           False |   4.135 ns | 0.0530 ns | 0.0470 ns |      - |         - |
|              SourceGeneratorEnabled |           False |  20.009 ns | 0.3119 ns | 0.2765 ns |      - |         - |
|             SourceGeneratorDisabled |           False |   3.208 ns | 0.0316 ns | 0.0296 ns |      - |         - |
|  SourceGeneratorDynamicLevelEnabled |           False |  20.603 ns | 0.2613 ns | 0.2316 ns | 0.0001 |      64 B |
| SourceGeneratorDynamicLevelDisabled |           False |   4.256 ns | 0.0445 ns | 0.0416 ns |      - |         - |
|                 **InterpolatedEnabled** |            **True** | **264.028 ns** | **2.9886 ns** | **2.6493 ns** | **0.0014** |     **664 B** |
|                InterpolatedDisabled |            True | 249.285 ns | 1.5408 ns | 1.3659 ns | 0.0014 |     664 B |
|                   DirectCallEnabled |            True | 354.462 ns | 2.4157 ns | 2.1414 ns | 0.0019 |     728 B |
|                  DirectCallDisabled |            True |  36.584 ns | 0.3609 ns | 0.3376 ns | 0.0001 |      64 B |
|               IsEnabledCheckEnabled |            True | 357.311 ns | 4.9134 ns | 4.5960 ns | 0.0019 |     728 B |
|              IsEnabledCheckDisabled |            True |   3.615 ns | 0.0750 ns | 0.0702 ns |      - |         - |
|          LoggerMessageDefineEnabled |            True | 340.671 ns | 4.9787 ns | 4.6570 ns | 0.0014 |     688 B |
|         LoggerMessageDefineDisabled |            True |   4.197 ns | 0.0887 ns | 0.0829 ns |      - |         - |
|              SourceGeneratorEnabled |            True | 331.744 ns | 2.5178 ns | 2.3552 ns | 0.0014 |     688 B |
|             SourceGeneratorDisabled |            True |   3.657 ns | 0.0644 ns | 0.0602 ns |      - |         - |
|  SourceGeneratorDynamicLevelEnabled |            True | 265.599 ns | 2.8241 ns | 2.6417 ns | 0.0019 |     728 B |
| SourceGeneratorDynamicLevelDisabled |            True |   4.243 ns | 0.0449 ns | 0.0398 ns |      - |         - |

[a comment by an automated account is not shown]
